content/blog/2019-10-30-loop-reduction/index.md

```diff
@@ -0,0 +1,375 @@
+
+
+
+# Challenges
+
+1. There are more properties we need than we originally expected. At first, we only generated loops, reaching variables. Then we for loop invariants code motion, we needed exits to the loop, dominance relations, live variables. 
+2. The representation of different variables are randomly decided at first and need implementation after we finalized the representation. For example,the loop invariant at first was stored as the list of instructions. Later, we found it necessary to change the storage form to the dictionary whose key is the block name. Otherwise, we would need to search and match the instruction to block, e.g, in the `move_LI` function.
+3. There are more similarity between basic induction variables and their families. It is sometimes tricky to differentiate them. Thus, the definition flow of each induction variable is maintained to tell them apart.
+
+
+
+## Evaluation
+
+1. We would like to evaluate our optimizer in two aspects: theoretical improvement and actual wall-clock speedup.
+
+   ### Benchmarks
+
+   Because loop invariant code motion and strength reduction are specific optimization for loops, and studies are based on case by case, we can hardly generate automatic testing scripts while writing bril codes to generate `for` loop are actually painful. Instead we reply on Typescript for help. 
+
+   However, default Bril scripts generated by Typescript always introduces redundancy during translation, and code motion optimization by default will be performed. For example, `int v10 = const 5; int v11 = v11 + v10;` can be found in the body of a loop if there is a line `x = x + 5` in Typescript. Because we don't perform other optimizations, we wonder how much speedup is contributed by the redundancy. We therefore write test cases with no expected optimizations (but will have some when translated to Bril) as baseline to compare actual wall-clock speedup.
+
+   | Filename                                | Description                                 |
+   | --------------------------------------- | ------------------------------------------- |
+   | `normal.ts`, `nested.ts`, `nestedif.ts` | No optimization expected.                   |
+   | `codemotion*.ts`                        | Expecting code motion.                      |
+   | `strengthreduction*.ts`                 | Expecting strength reduction optimizations. |
+   | `both*.ts`                              | Expecting both kinds of optimizations.      |
+
+
+
+   ### Implicit Redundancy
+
+   Bril scripts generated by Typescript introduce some redundant operations usually an `id` or a `const`. Inside a loop, these would usually be loop invariant code which gets removed by our optimizer. However, we were not sure if this would contribute towards outr speedup. To make sure that these are actually negligible we benchmarked similar codes without explicit loop invariant codes in the loop to see speedups (if any).
```

Comment:
"some redundant operations: usually" (insert some sort of punctuation; probably a colon)
"outr" (please run spell check)